flocker/node/agents/test/test_blockdevice.py

```diff
@@ -0,0 +1,459 @@
+# Copyright Hybrid Logic Ltd.  See LICENSE file for details.
+
+"""
+Tests for ``flocker.node.agents.blockdevice``.
+"""
+
+import os
+from uuid import uuid4
+from operator import attrgetter
+from subprocess import check_output
+
+from zope.interface.verify import verifyObject
+
+from ..blockdevice import (
+    LoopbackBlockDeviceAPI, IBlockDeviceAPI,
+    BlockDeviceVolume, UnknownVolume, AlreadyAttachedVolume,
+    losetup_list_parse, losetup_list, UnattachedVolume
+)
+
+from twisted.python.filepath import FilePath
+from twisted.trial.unittest import SynchronousTestCase, SkipTest
+
+
+def sorted_by_blockdevice_id(blockdevices):
+    """
+    :param list blockdevices: The ``BlockDeviceVolume`` instances to sort.
+    :returns: A ``list`` of ``BlockDeviceVolume``s sorted by their
+        ``blockdevice_id``.
+    """
+    return sorted(blockdevices, key=attrgetter('blockdevice_id'))
+
+
+class IBlockDeviceAPITestsMixin(object):
+    """
+    Tests to perform on ``IBlockDeviceAPI`` providers.
+    """
+    def test_interface(self):
+        """
+        ``api`` instances provide ``IBlockDeviceAPI``.
+        """
+        self.assertTrue(
+            verifyObject(IBlockDeviceAPI, self.api)
+        )
+
+    def test_list_volume_empty(self):
+        """
+        ``list_volumes`` returns an empty ``list`` if no block devices have
+        been created.
+        """
+        self.assertEqual([], self.api.list_volumes())
+
+    def test_created_is_listed(self):
+        """
+        ``create_volume`` returns a ``BlockVolume`` that is returned by
+        ``list_volumes``.
+        """
+        new_volume = self.api.create_volume(size=1024)
+        self.assertIn(new_volume, self.api.list_volumes())
+
+    def test_attach_unknown_volume(self):
+        """
+        An attempt to attach an unknown ``BlockDeviceVolume`` raises
+        ``UnknownVolume``.
+        """
+        self.assertRaises(
+            UnknownVolume,
+            self.api.attach_volume,
+            blockdevice_id=unicode(uuid4()),
+            host=b'192.0.2.123'
```

Comment:
I'll leave it for a followup issue, but won't create a separate issue. We'll fix this when we implement the first real block device backend. I added a FIXME comment though.